cmd/ebpf/pdr.go

```diff
@@ -16,26 +17,77 @@ type PdrInfo struct {
 	OuterHeaderRemoval uint8
 	FarId              uint32
 	QerId              uint32
+	SdfFilter          *SdfFilter
+}
+
+type SdfFilter struct {
+	Protocol     uint8 // 0: icmp, 1: ip, 2: tcp, 3: udp
+	SrcAddress   IpWMask
+	SrcPortRange PortRange
+	DstAddress   IpWMask
+	DstPortRange PortRange
+}
+
+type IpWMask struct {
+	Type uint8 // 0: any, 1: ip4, 2: ip6
+	Ip   net.IP
+	Mask net.IPMask
+}
+
+type PortRange struct {
+	LowerBound uint16
+	UpperBound uint16
+}
+
+func HandlePdrWithSdf(lookup func(interface{}, interface{}) error, key interface{}, pdrInfo PdrInfo) (*IpEntrypointPdrInfo, error) {
+	var pdrToStore IpEntrypointPdrInfo
+	if pdrInfo.SdfFilter != nil {
```

Comment:
Let's get this condition out of the function?

Reply:
To be honest, I don't quite understand an intention of the question.

Is that better?
```
func HandlePdrWithSdf(lookup func(interface{}, interface{}) error, key interface{}, pdrInfo PdrInfo) (IpEntrypointPdrInfo, error) {
	var defaultPdr IpEntrypointPdrInfo
	if err := lookup(key, &defaultPdr); err != nil {
		return nil, err
	}
	return CombinePdrWithSdf(defaultPdr, pdrInfo), nil
}

func (bpfObjects *BpfObjects) PutPdrUpLink(teid uint32, pdrInfo PdrInfo) error {
	log.Printf("EBPF: Put PDR Uplink: teid=%d, pdrInfo=%+v", teid, pdrInfo)
	var pdrToStore IpEntrypointPdrInfo
	var err error
	if pdrInfo.SdfFilter != nil {
		if pdrToStore, err = HandlePdrWithSdf(bpfObjects.PdrMapUplinkIp4.Lookup, teid, pdrInfo); err != nil {
			return err
		}
	} else {
		pdrToStore = ToIpEntrypointPdrInfo(pdrInfo)
	}
	return bpfObjects.PdrMapUplinkIp4.Put(teid, unsafe.Pointer(&pdrToStore))
}
```

Or that?
```
func HandlePdrWithSdf(lookup func(interface{}, interface{}) error, key interface{}, pdrInfo PdrInfo) (IpEntrypointPdrInfo, error) {
	var defaultPdr IpEntrypointPdrInfo
	if err := lookup(key, &defaultPdr); err != nil {
		return nil, err
	}
	return CombinePdrWithSdf(defaultPdr, pdrInfo), nil
}

func PreprocessPdr(lookup func(interface{}, interface{}) error, key interface{}, pdrInfo PdrInfo) (IpEntrypointPdrInfo, error) {
	if pdrInfo.SdfFilter != nil {
		return HandlePdrWithSdf(lookup, key, pdrInfo)
	} else {
		return ToIpEntrypointPdrInfo(pdrInfo), nil
	}
}

func (bpfObjects *BpfObjects) PutPdrUpLink(teid uint32, pdrInfo PdrInfo) error {
	log.Printf("EBPF: Put PDR Uplink: teid=%d, pdrInfo=%+v", teid, pdrInfo)
	if pdrToStore, err := HandlePdrWithSdf(bpfObjects.PdrMapUplinkIp4.Lookup, teid, pdrInfo); err == nil {
		return bpfObjects.PdrMapUplinkIp4.Put(teid, unsafe.Pointer(&pdrToStore))
	} else {
		return err
	}
}
```

Or neither?

My original idea of creating function HandlePdrWithSdf was reducing the code size in functions like PutPdrUpLink, PutPdrDownLink, etc.

Reply:
The intention is to show that only PDRs with SDF require additional handling. That's why `if pdrInfo.SdfFilter != nil` in root function would look better.